쥬스 메이커 [STEP1] Groot, borysarang

Groot-94 opened the pull request:
@protocorn93 
안녕하세요 콘! 
첫 PR 보냅니다 :) 잘 부탁드립니다 ㅎㅎ 🫡

---
### 함수기능
- takeOrder : 주문을 받고 재고를 확인해서 쥬스를 만드는 함수
- usedLeftover : 쥬스를 만들기 위해 재고를 줄이는 함수
- checkInventory : 쥬스를 만들 수 있는지 재고를 확인하는 함수
- checkGenerationAvailable : 레시피에 따른 재료를 확인하고 checkInventory를 호출하는 함수
- generate : 레시피에 따른 재료를 확인하고 usedLeftover를 호출하는 함수
- make : checkInventory, checkGenerationAvailable 를 외부에서 호출 할 수 있도록 캡슐화 한 함수
- getRecipe : 쥬스에 따른 레시피를 반환하는 함수

### 고민한점 🥹
- 은닉화 : FruitStore 클래스의 메서드들을 internal로 설정하면 FruitStore 객체를 만들어서 과일의 재고만 변동을 줄 수도 있다는 생각이 들었습니다. 그래서 FruitStore의 메서드를 JuiceMaker 에서만 사용하게 하고 싶었으나 은닉화를 하게 되면 JuiceMaker에서 사용할 수 없어서 `어떻게 해야 하는 게 좋은지`, `이 생각을 하는게 옳은 것인지 궁금합니다`. 
    - 해결방법 : FruitStore에서 은닉화가 필요한 함수를 은닉화 해주고, 그 함수를 다른 함수로 묶어서 쥬스를 만들지 않으면 과일의 재고를 변경 할 수 없도록 해줬습니다. 그리고 JuiceMaker의 takeOrder 함수에선 쥬스를 만드는 함수 make만 호출하도록 수정했습니다.

- Dictionary : 처음에 주스레시피를 Dictionary으로 관리를 하다 보니 Dictionary의 값을 옵셔널 바인딩해줘야 했습니다. 그렇게 되면 레시피가 두 가지의 과일을 사용하는 경우에는(딸기 바나나주스) for문을 사용해야 했고, for문 안에서 옵셔널 바인딩을 해야 했기 때문에 이중 들여쓰기를 사용해야 했고 이중들여쓰기 사용은 프로젝트 규칙을 어기는 방법이였습니다. 
    - 해결방법 : FruitJuice 열거형에서 각 레시피에 맞는 과일과 수량을 튜플에 넣고 그 `튜플을 배열에 넣어 리턴하는 방식`으로 사용했습니다.

Groot-94 committed:
docs: README.md 양식 추가

Groot-94 committed:
Create README.md

yusw10 committed:
add: Create FruitJuice Enum

yusw10 committed:
Merge branch 'step1' of https://github.com/Groot-94/ios-juice-maker into step1

# Conflicts:
#	README.md

Groot-94 committed:
add: Create JuiceMakerError Enum

yusw10 committed:
add: Create Fruit Enum

Groot-94 committed:
feat: Create FruitStore in property, function

- fruitInventory property
- changeInventory function
- checkGenerationAvailable function

yusw10 committed:
feat: Create JuiceMaker in property, function

- fruitStore property
- make function
- printError function

Groot-94 committed:
feat: Implement changeInventory function

yusw10 committed:
feat: Implement checkGenerationAvailable function of FruitStore

Groot-94 committed:
feat: Implement make function, checkAbout function of JuiceMaker

yusw10 committed:
refactor: Refactoring checkAbout function

 - make getRecipe function to FruitJuice enum
 - refactoring checkAbout Function
 - add debug code to Viewcontroller

Groot-94 committed:
feat: Create generate function to JuiceMaker

- checkAbout function naming update to checkGenerationAvailable
- changeInventory function naming update to usedLeftover
- checkGenerationAvailable function naming update to checkInventory

yusw10 committed:
refactor : Refactoring function and property

 - hiding checkGenerationAvailable function
 - hiding generate function
 - replace magic number with symbolic constant

Groot-94 committed:
refactor: Refactoring function hiding, naming

- move function checkGenerationAvailable, generate to FruitStore from juiceMaker
- update naming generate -> make / make -> takeOrder

Groot-94 committed:
docs: Update README.md

Groot-94 committed:
docs : Update README.md

protocorn93 commented on JuiceMaker/JuiceMaker/Model/JuiceMakerError.swift:
```diff
@@ -0,0 +1,11 @@
+
+enum JuiceMakerError: Error {
+    case outOfStock
+    case invalidOder
```
invalidOder > invalidOrder

오타 수정이 필요해보여요

yusw10 replied:
수정완료 했습니다..!

protocorn93 replied:
확인했어요! 한 가지 PR 커뮤니케이션 방법을 말씀드리자면 수정을 요청하거나 어떤 피드백에 대한 수정 커밋이 발생했다면 해당 코멘트에 함께 남겨주시면 리뷰어가 보다 편하게 확인할 수 있어요!

> [801b6e4](https://github.com/yagom-academy/ios-juice-maker/pull/222/commits/801b6e4215c377525246bb543da4d3caf34366af) 여기에 수정했어요~

처럼요!

Groot-94 replied:
넵 다음부턴 사용해보겠습니다!

protocorn93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -8,5 +8,47 @@ import Foundation
+    private var fruitInventory: [Fruit:Int] = [:]
     
+    init() {
+        let initialAmount: Int = 10
```
기본 기능 명세가 10개이기에, 여기에 갯수를 정의한 것 같아요. 하지만 initialAmount를 생성자로 받는 것은 어떨까요? 개인적으론 이런 방법을 선호하는 편이에요. 지금은 단순히 “10개”라는 것이지만 저는 이것도 일종의 명세로 언제든 바뀔 수 있는 것이라고 생각해요. 바뀔 수 있는 것은 외부에서 넣어주어야 변경에 보다 유연하게 대처할 수 있다고 생각해요. 

맞아요. 지금 당장은 이게 어떤 말인지 이해가 안가실 수 있어요. 지금은 단순히 “자주 변경될 수 있는 것과 거의 변경되지 않는 것”을 분리하는 연습을 한다고 생각해주시면 되요. 그 둘의 기준도 여러분들이 정해보시면 되요. 지금은 너무 깊게 생각하시지 않아도 되요. 이런 개념들은 추후 프로젝트를 진행하면서 꾸준히 보고 공부하셔야 할 개념들이에요!

Groot-94 replied:
수정했습니다.
“자주 변경될 수 있는 것과 거의 변경되지 않는 것” 명심하겠습니다!

yusw10 replied:
오늘(04.28) 야곰과 활동학습으로 프로퍼티와 이니셜라이저의 역할에 대해 이야기 해보았습니다. 코멘트 달아주신 것처럼 10개라는것은 이번 STEP의 명세일 뿐이고 만약 사용자로부터 각각의 재고를 입력받는 형태 였다면 의미에 부합하지 않는 형태인것 같습니다. 
그래서 각 과일의 초기 재고량도 재각각 다를 수도 있다고 생각하지만, 이번 STEP에서는 모두가 10개라고 지정되어 있어서 이니셜라이저로 동일한 수량을 받도록 수정해보았습니다.

protocorn93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -8,5 +8,47 @@ import Foundation
+    }
+    
+    private func usedLeftover(_ fruit: Fruit, by amount :Int) {
+        guard (fruitInventory[fruit]? -= amount) != nil else {
```
개인적으로는 가독성이 조금은 떨어지는 것 같아요. 
- 키에 대한 값이 존재하는지 여부 
- 충분한 수량을 갖고 있는지 여부

이렇게 두개의 guard 문으로 나눠보시는건 어떨까요?

Groot-94 replied:
기존에 checkInventory 함수에서 충분한 수량을 갖고 있는지 여부를 확인하고 usedLeftover를 호출하는 방향을 생각했었습니다. 
 하지만, usedLeftover에서도 키에 대한 값이 존재하는지 여부를 확인해야 수량을 줄일 수 있으니 결국엔 확인을 두번하는? 느낌을 받았습니다. 그래서 두 함수를 합치고 함수안에서 guard 문을 사용해 `키에 대한 값이 존재하는지 여부를 확인` -> `그 값이 필요한 수량보다 많은지 확인` -> `필요한 수량을 재고에서 차감` 의 순서로 수정했습니다.

protocorn93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -8,5 +8,47 @@ import Foundation
+        print("남은 \(fruit)의 수: \(fruitInventory[fruit] ?? 0)")
+    }
+    
+    private func checkInventory(about fruit: Fruit, by amount :Int) throws {
```
보통 `check`와 같이 여부를 검사하는 메서드는 bool 값을 반환해주는 것이 좀 더 일반적이라고 생각해요. 더 명확한 의미를 전달해줄 수 있구요. 

에러를 언제 던지는게 좋을지에 대해서도 함께 고민해보면 좋을 것 같아요.

protocorn93 replied:
https://docs.swift.org/swift-book/LanguageGuide/ErrorHandling.html

yusw10 replied:
위의 guard를 처리하는 부분과 여기서 리턴을 해주는 부분에 대해서 그루트와 많이 이야기를 나눠보았습니다. 
우선 프로젝트 정의에 "(Optional) 필요한 경우 오류처리(throw, try-catch)를 구현합니다" 라고 기입이 되어 있었고 저희는 이 부분이 optional을 해제할 때 throw 및 try-catch로 처리하라는 의미로 이해했습니다. 

작성한 guard문이 동작하는 경우는 
1. 요청한 음료가 없는 음료일 경우(메뉴에 없는, optional 바인딩을 하기위한.)
2. 음료 제작에 필요한 과일의 수량이 모자라는 경우

이렇게 두가지가 사용이 되고 있었고 저희의 코드에서는 이 부분을 모두 Error 프로토콜을 상속받아 try catch에서 각각 잡아주도록 만들었습니다.
이후에 코멘트를 다시 읽어보니 **에러**가 어떤 상황에 나는지 생각해보았습니다. 저희는 사용자가 프로그램을 사용할 때
1. 제작자가 의도 하지 않은 동작을 할 경우
2. 런타임 에러를 비롯한 개발자가 확인하지 못한 동작 불량

이 두가지 경우만 에러라고 정의해보았습니다. 
따라서 "재고가 없는 경우 사용자가 정상적으로 요청을 했지만 이를 수행할 수 없는경우"는 에러가 아니라고 생각했기에 코드의 재고확인하는 부분중 throw를 제거하고 `Bool`타입으로 변경해보았습니다.

protocorn93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -8,5 +8,47 @@ import Foundation
+        }
+    }
+    
+    private func generate(_ fruitJuice: FruitJuice) {
```
마찬가지로 결과를 반환해주는게 좋아요. 우리가 무언가를 검증할 땐 input 대비 output을 검증하잖아요? 프로그램도 마찬가지로 input에 대한 output으로 검증을 할 수 있도록 만드는 것이 좋아요. 

이는 추후에 여러분이 학습하게 될 단위 테스트에 대해 학습할 때 도움이 되실거에요.

Groot-94 replied:
리뷰를 보면서 input, output을 검증하는 방향으로 코드를 수정해봤습니다
항상 input은 고려했지만 output은 고려하지 않았던 것 같습니다. 아직까지는 " 이 함수에선 어떤 output이 나오겠구나"라고 생각만 하고 넘어가도 문제가 없었지만, 나중에 큰 프로젝트나 다양한 output이 나오는 상황에서는 문제가 될 수도 있겠다는 생각이 듭니다. 감사합니다!

protocorn93 commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -8,5 +8,47 @@ import Foundation
+        }
+    }
+    
+    func make(_ fruitJuice: FruitJuice) throws {
```
generate 메서드와 동일한 코멘트입니다.

protocorn93 reviewed (changes requested):
두 분 모두 수고하셨어요! 훌륭하게 코드 작성해주셨네요! 👍 UML까지!! 👍 

몇 가지 소소한 코멘트 남겨놨어요. 읽어보시고 반영해주실 수 있는 것들은 반영해주시면 좋을 것 같아요.

### 고민한점 
- 은닉화 
  - 많이 고민해보셔야 할 부분인데 유의미한 시간을 가지셨네요! 은닉화는 매번 고민이죠! 은닉화도 그 밑바탕에는 결국 객체들의 역할이 적절하게 나뉘어져 있는지에요. 단순히 숨겨야지! 라는 생각보단 이 객체에겐 어떤걸 요청해야할까? 라는 고민을 해보시면 좋을 것 같아요. 그리고 요청을 받은 객체는 어떻게 응답을 해야할까? 에 대해 고민해보시고 은닉화는 여기서 “어떻게”를 숨기는 것이라고 생각하시면 되요! 

- Dictionary
  - 튜플도 이 프로젝트에선 충분히 활용해볼 수 있는 요소라고 생각해요. 그리고 적절한 선택이었다고 생각해요.
  - 하지만 들여쓰기를 하나만 쓸 수 있는 가장 좋은 방법은 적절하고 작은 단위로 메서드를 나누는 것이에요. 이 점도 생각해보시면서 코드를 작성해보시면 좋을 것 같아요!

yusw10 committed:
refactor: Refactoring initializer to initialAmount

 - refactoring init() function
 - make initalAmount property

Groot-94 committed:
refactor: Refactor update function of FruitStore, update function of JuiceMaker

FruitStore
- update usedLeftover function
- remove checkGenerationAvailable function
- remove generate function
- update make function

JuiceMaker
- update  takeOrder function

yusw10 committed:
refactor: Refactoring function

 - refactoring error handling in usedLeftOver
 - rename function name usedLeftOver to canUseStock
 - refactoring make function -> return Optional value
 - refactoring JuiceMakerError case

Groot-94 commented:
@protocorn93
안녕하세요 콘!
두번째 PR 보냅니다 :) 잘 부탁드립니다 ㅎㅎ 🫡

protocorn93 reviewed (approved):
제가 남긴 코멘트에 대해서 단순히 수정하는 것에 그치지 않고, 여러분들이 생각을 해본 방향성에 대해 말씀해주셔서 좋았어요! 

앞으로도 이렇게 피드백에 대해 단순히 수용의 코멘트 뿐만 아니라 여러분들이 작성한 코드에 대한 생각을 함께 적어주시면 좋을 것 같아요! 수고하셨어요! 

그리고 UML도 PR에 함께 첨부해주시면 좋을 것 같아요!
